File: hermes-trading-post/test_websocket.py
#!/usr/bin/env python3
"""Test Coinbase WebSocket connection"""
import websocket
import json
import threading
import time

price_received = False
current_price = 0

def on_message(ws, message):
    global price_received, current_price
    try:
        data = json.loads(message)
        if data.get('type') == 'ticker' and data.get('product_id') == 'BTC-USD':
            price = float(data.get('price', 0))
            current_price = price
            price_received = True
            print(f"✅ Price received: ${price:,.2f}")
    except Exception as e:
        print(f"❌ Error: {e}")

def on_open(ws):
    print("🔌 WebSocket connected")
    subscribe_msg = {
        "type": "subscribe",
        "product_ids": ["BTC-USD"],
        "channels": ["ticker"]
    }
    ws.send(json.dumps(subscribe_msg))
    print("📤 Subscription sent")

def on_error(ws, error):
    print(f"❌ WebSocket error: {error}")

def on_close(ws, close_status_code, close_msg):
    print(f"🔌 WebSocket closed: {close_status_code} - {close_msg}")

print("🧪 Testing Coinbase WebSocket...")
print("=" * 50)

try:
    ws = websocket.WebSocketApp(
        "wss://ws-feed.exchange.coinbase.com",
        on_message=on_message,
        on_error=on_error,
        on_close=on_close,
        on_open=on_open
    )
    
    # Run in background thread
    def run_ws():
        ws.run_forever()
    
    ws_thread = threading.Thread(target=run_ws, daemon=True)
    ws_thread.start()
    
    # Wait for price
    print("⏳ Waiting for price data...")
    timeout = 10
    start = time.time()
    
    while not price_received and (time.time() - start) < timeout:
        time.sleep(0.1)
    
    if price_received:
        print(f"\n✅ SUCCESS! WebSocket is working")
        print(f"   Current BTC price: ${current_price:,.2f}")
    else:
        print(f"\n❌ TIMEOUT: No price received after {timeout} seconds")
        print("   Check your internet connection")
    
    ws.close()
    
except Exception as e:
    print(f"❌ Failed to connect: {e}")
    print("   Check if websocket-client is installed: pip install websocket-client")

File: hermes-trading-post/test-websocket.js
import WebSocket from 'ws';

console.log('Testing Coinbase WebSocket with heartbeat...');

const ws = new WebSocket('wss://ws-feed.exchange.coinbase.com');
let messageCount = 0;
let lastMessageTime = Date.now();

ws.on('open', () => {
  console.log('WebSocket connected');
  
  // Subscribe to BTC-USD ticker
  const subscribeMessage = {
    type: 'subscribe',
    product_ids: ['BTC-USD'],
    channels: ['ticker']
  };
  
  ws.send(JSON.stringify(subscribeMessage));
  console.log('Sent subscribe message');
  
  // Start heartbeat - resubscribe every 30 seconds
  setInterval(() => {
    if (ws.readyState === WebSocket.OPEN) {
      const timeSinceLastMessage = Date.now() - lastMessageTime;
      console.log(`\n[Heartbeat] Time since last message: ${Math.round(timeSinceLastMessage/1000)}s`);
      console.log('[Heartbeat] Re-subscribing to keep connection alive...');
      ws.send(JSON.stringify(subscribeMessage));
    }
  }, 30000);
});

ws.on('message', (data) => {
  messageCount++;
  lastMessageTime = Date.now();
  const message = JSON.parse(data.toString());
  
  if (message.type === 'subscriptions') {
    console.log('\nSubscription confirmed:', message);
  } else if (message.type === 'ticker' && message.product_id === 'BTC-USD') {
    console.log(`\n[${new Date().toISOString()}] BTC Price: $${message.price} (message #${messageCount})`);
  } else if (message.type === 'error') {
    console.error('\nError message:', message);
  }
});

ws.on('close', () => {
  console.log('\nWebSocket disconnected');
  process.exit(0);
});

ws.on('error', (error) => {
  console.error('\nWebSocket error:', error);
});

// Run for 2 minutes to test heartbeat
setTimeout(() => {
  console.log(`\nTest complete. Received ${messageCount} messages.`);
  ws.close();
}, 120000);

console.log('Test will run for 2 minutes to verify heartbeat functionality...');